apps/provider-proxy/src/services/ProviderProxy.ts

```diff
@@ -1,24 +1,151 @@
+import { X509Certificate } from "crypto";
 import https, { RequestOptions } from "https";
-import fetch, { BodyInit, Headers, Response } from "node-fetch";
+import { LRUCache } from "lru-cache";
+import { text } from "stream/consumers";
+import { PeerCertificate, TLSSocket } from "tls";
+
+import { validateCert } from "../utils/validateCertificates";
 
 export class ProviderProxy {
+  private readonly knownCertificatesCache = new LRUCache<string, Record<string, PeerCertificate>>({
+    max: 1000,
+    ttl: 30 * 60 * 1000,
+    allowStale: true
+  });
+
+  constructor(private readonly now: () => number) {}
+
   fetch(url: string, options: FetchOptions): Promise<Response> {
-    const httpsAgent = new https.Agent({
-      cert: options.cert,
-      key: options.key,
-      rejectUnauthorized: false
-    });
+    return new Promise<Response>((resolve, reject) => {
+      const req = https.request(
+        url,
+        {
+          method: options.method,
+          headers: {
+            "Content-Type": "application/json",
+            ...options.headers
+          },
+          cert: options.cert,
+          key: options.key,
+          rejectUnauthorized: false
+        },
+        async res => {
+          try {
+            res.setEncoding("utf8");
+            const socket = res.socket as TLSSocket;
+            const serverCert = socket.getPeerCertificate();
+            // @see https://nodejs.org/api/tls.html#session-resumption
+            // for servers which support TLS session resumption, handshake phase is skipped for subsequent requests
+            // to improve performance and in this case certicate is not available because it is not requested.
+            // There is a way to disable session resumption but it will hurt performance.
+            // To disable either create a new `https.Agent` for every request or reduce sesion related options in it
+            // sessionTimeout & maxCachedSessions. In this case, we will do TLS handshake on every request and
+            // will have certificate for every request
+            const didHandshake = Object.keys(serverCert).length > 0;
+
+            if (didHandshake && options.network && options.providerAddress) {
```

Comment:
temporary ignore validation if `options.network` and `options.providerAddress` are empty because `deploy-web` also require adjustments, just didn't want to create massive PR